adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,280 @@
+
+Non-API packages should be under `.impl` or `internal` like `io.quarkus.<extension-name>.runtime.internal` or `io.quarkus.<extension-name>.runtime.impl`.
+Thus, automated tooling can easily use this information to help generate a `module-info` with the correct exports, as well as excluding those packages from JavaDoc generation.
+
+==== `build` module
+* `io.<quarkus|quarkiverse>.<extension-name>.build`: Internal build logic (processors, build steps, internal build items). Public SPIs must reside in the `build-spi` module. May include subpackages (excluding `spi`).
+
+==== `build-spi` module
+* `io.<quarkus|quarkiverse>.<extension-name>.build.spi`: Public build items and types. Considered public API and subject to compatibility guarantees.
+
+Build items in this module should be documented with their intended use (produced or consumed).
+
+==== `runtime-api` module
+* `io.<quarkus|quarkiverse>.<extension-name>.api`: Public runtime API. Can be used independently of the full extension. Consumers should not expect the full extension to be available at runtime. To make sure the extension is available, the consumer should use the `runtime` module (which would pull in the `runtime-api` module transitively).
+
+==== `runtime-dev` module
+* `io.<quarkus|quarkiverse>.<extension-name>.dev`: Dev-mode-only runtime classes, e.g., for Dev UI contribution. Not included in production builds.
+
+==== `codegen` module
+
+* `io.<quarkus|quarkiverse>.<extension-name>.codegen`: Code generation logic, if applicable. This module is optional and may not be present in all extensions.
+
+
+=== Module Summary Table
+
+[cols="1,3,2",options=“header"]
+|===
+| Module | Purpose | Intended Consumers
+
+| runtime
+| Runtime logic and extension internals
+| Application code, Quarkus runtime
+
+| runtime-api
+| Public runtime APIs and service provider types
+| Other extensions, libraries
+
+| build
+| Build steps, processor logic, internal build items
+| Quarkus build system
+
+| build-spi
+| Shared build-time APIs (build items, metadata)
+| Other extensions
+
+| runtime-dev
+| Dev mode–only logic (e.g., Dev UI contributions)
+| Development-time only, not production
+
+| codegen
+| Code generation logic (if applicable)
+| Quarkus build system
+
+| cli
+| `quarkus` CLI plugin
+| The Quarkus CLI
+|===
+
+== Consequences
+
+=== Positive
+
+	* Improves long-term maintainability and clarity of the codebase.
+	* Clarifies the public API surface and encourages proper separation of concerns.
+	* Avoids the creation of split packages and internal dependency leakage.
+	* Lays the groundwork for potential future adoption of JPMS (Java Platform Module System).
+
+=== Negative
+
+	* Adds structural complexity, which may feel unnecessary for simple extensions.
+	* Refactoring existing extensions to adopt this structure requires engineering effort.
```

Comment:
Have we estimated how much time it will take to migrate an average sized extension to understand if we really can commit to this?
Is there a prototype somewhere?

Reply:
I think you can automate part of it but not all of it as the current status is not great and the reason why we have to split things is that our code structure is not completely clear.

It's going to be a massive amount of work and quite error prone.

And we will need to be extremely careful for some modules, especially the core one.

For me, we need to define a target first, try with the core extension and a few extensions and check the consequences of these changes on our codebase.